fix: Add pagination to Potential Duplicate Get Query [TECH-644]

lucaCambi77 opened the pull request:
https://dhis2.atlassian.net/browse/TECH-644

This issue was initially planned with the Deduplication feature. It adds a quick implementation for the Potential Duplicate pagination.
Actually, in the front-end (both old and new Capture) ) the Potential Duplicate table is not used at all, so in https://dhis2.atlassian.net/browse/DHIS2-14599 the front-end team will see how this can be further improved or adapted if required at some point. If not, it is worth considering deprecating the `/potentialDuplicates` endpoint also from the backend

enricocolasante commented on dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/deduplication/DeduplicationController.java:
```diff
@@ -159,18 +151,18 @@ public PotentialDuplicate postPotentialDuplicate(
         throws NotFoundException,
         BadRequestException
     {
         checkDeduplicationStatusRequestParam( status );
```
This check will never throw an Exception. If an invalid value is passed for the DeduplicationStatus then Spring will throw a `MethodArgumentTypeMismatchException` that will be intercepted in `CrudControllerAdvice` and translated to a `BadRequestException`

enricocolasante commented on dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/deduplication/DeduplicationController.java:
```diff
@@ -97,40 +92,37 @@
     {
-        checkDeduplicationStatusRequestParam(
-            Optional.ofNullable( query.getStatus() ).map( Enum::name ).orElse( null ) );
+        checkDeduplicationStatusRequestParam( potentialDuplicateCriteria.getStatus() );
```
This check will never throw an Exception. If an invalid value is passed for the DeduplicationStatus then Spring will throw a `BindingException` that should be intercepted in CrudControllerAdvice and translated to a BadRequestException. The interceptor for this exception is not implemented yet

enricocolasante commented on dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/deduplication/DeduplicationController.java:
```diff
@@ -229,11 +221,11 @@ private void checkDbAndRequestStatus( PotentialDuplicate potentialDuplicate,
     }
 
-    private void checkDeduplicationStatusRequestParam( String status )
+    private void checkDeduplicationStatusRequestParam( DeduplicationStatus status )
```
This shouldn't be necessary then

lucaCambi77 replied:
yes this is old code so I need to review with your suggestion

enricocolasante reviewed (commented):
I think there is something odd in the Controller and the validation of the `DeduplicationStatus`.
Generally is good to have a test in `web-api-test` to have all the test cases to cover the different invalid parameters and the errors returned

lucaCambi77 commented:
@enricocolasante I've updated MVC and added integration tests. There is should be still one test failing  related to https://dhis2.atlassian.net/browse/DHIS2-13928

teleivo commented on dhis-2/dhis-web-api-test/src/test/java/org/hisp/dhis/webapi/controller/deduplication/DeduplicationControllerTest.java:
```diff
@@ -0,0 +1,169 @@
+    void shouldThrowBindExceptionWhenGetPotentialDuplicateWithInvalidStatus()
+    {
+        assertStatus( HttpStatus.BAD_REQUEST,
+            GET( ENDPOINT + "?status=" + "invalidStatus" ) );
```
@enricocolasante should we test these BindExceptions and MethodArgumentTypeMismatchExceptions in individual controllers? Or is it simply duplicating tests you already have elsewhere?

teleivo commented on dhis-2/dhis-web-api-test/src/test/java/org/hisp/dhis/webapi/controller/deduplication/DeduplicationControllerTest.java:
```diff
@@ -0,0 +1,169 @@
+        throws Exception
+    {
+        PotentialDuplicate potentialDuplicate = new PotentialDuplicate( origin.getUid(), null );
+        assertStatus( HttpStatus.BAD_REQUEST, POST( ENDPOINT, objectMapper.writeValueAsString( potentialDuplicate ) ) );
```
We have multiple tests that are testing the same endpoint and method. Since we only assert on the status code there is a risk that multiple tests test the same case. Even if this is not the case now it could be after a refactoring and not be caught by the tests.

I wonder if there is a keyword specific to these cases that we can assert on in the body? Not asserting on the exact message.

teleivo commented on dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/deduplication/hibernate/HibernatePotentialDuplicateStore.java:
```diff
@@ -104,64 +109,79 @@ public HibernatePotentialDuplicateStore( SessionFactory sessionFactory, JdbcTemp
 
-            return hibernateQuery.getSingleResult().intValue();
-        } ).orElseGet( () -> {
+        Predicate[] predArray = getQueryPredicates( query, cb, root );
```
I would inline vars that are directly used. Less to read 😄

teleivo commented on dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/deduplication/hibernate/HibernatePotentialDuplicateStore.java:
```diff
@@ -104,64 +109,79 @@ public HibernatePotentialDuplicateStore( SessionFactory sessionFactory, JdbcTemp
+                    root.get( "duplicate" ).in( query.getTeis() ) ) ) );
         }
+
+        Predicate[] predicates = new Predicate[predicateList.size()];
```
Is intellij flagging this for you? 🤔 

According to https://stackoverflow.com/questions/9572795/convert-list-to-array-in-java

`predicateList.toArray(new Predicate[0]);`

should be preferred

teleivo commented on dhis-2/dhis-test-integration/src/test/java/org/hisp/dhis/deduplication/DeduplicationServiceIntegrationTest.java:
```diff
@@ -71,24 +73,23 @@ public void setUpTest()
-        assertEquals( 2, deduplicationService.getAllPotentialDuplicatesBy( potentialDuplicateQuery ).size() );
+
+        criteria.setStatus( DeduplicationStatus.OPEN );
+        assertEquals( 1, deduplicationService.getPotentialDuplicates( criteria ).size() );
```
Such assertions especially in integration tests can be frustrating once the test is flaky and only fails on CI. They just tell us `expected 1, got 3` if the sizes don't match. Ideally we also see whats in the collection if sizes don't match.

What do you think of adding and using

```java
    /**
     * Asserts that the given collection is of the expected size. Use {@link #assertIsEmpty(Collection)} instead of asserting on size 0.
     *
     * @param expectedSize the expected size of actual
     * @param actual the actual collection to check
     */
    public static <T> void assertSize(int expectedSize, Collection<T> actual) {
        assertNotEmpty(actual);
        assertEquals(expectedSize, actual.size(), () -> String.format("Expected collection to contain %d elements, got %d instead: %s", expectedSize, actual.size(), actual));
    }
```

here https://github.com/dhis2/dhis2-core/blob/ecf7d9990ec39aaeded74cd287633022c7769493/dhis-2/dhis-support/dhis-support-test/src/main/java/org/hisp/dhis/utils/Assertions.java#L112

gives us a message like

```
org.opentest4j.AssertionFailedError: Expected collection to contain 6 elements, got 3 instead: [1, 2, 3] ==> 
Expected :6
Actual   :3
```

teleivo reviewed (approved):
Great 😄 
Some minor comments and a question for @enricocolasante

enricocolasante reviewed (approved):
Just a small comment on redundant tests
